diff -up virt-who-0.16/virtwho.py.fix virt-who-0.16/virtwho.py
--- virt-who-0.16/virtwho.py.fix	2016-03-10 16:57:06.676109252 +0100
+++ virt-who-0.16/virtwho.py	2016-03-10 16:57:33.234898758 +0100
@@ -743,8 +743,7 @@ def main():
         sd_notify("READY=1\nMAINPID=%d" % os.getpid())
         while True:
             try:
-                _main(virtWho)
-                break
+                return _main(virtWho)
             except ReloadRequest:
                 logger.info("Reloading")
                 continue
@@ -764,9 +763,12 @@ def _main(virtWho):
                 if report == 'reload':
                     raise ReloadRequest()
                 elif report == 'exit':
-                    return
+                    return 0
 
     if virtWho.options.print_:
+        if len(result) == 0:
+            virtWho.logger.error("No hypervisor reports found")
+            return 1
         hypervisors = []
         for config, report in result.items():
             if isinstance(report, DomainListReport):
@@ -785,8 +787,9 @@ def _main(virtWho):
         })
         virtWho.logger.debug("Associations found: %s", json.dumps({
             'hypervisors': hypervisors
-        }, indent=4))
+        }, indent=4, sort_keys=True))
         print(data)
+    return 0
 
 
 def exit(code, status=None):
@@ -807,7 +810,7 @@ def exit(code, status=None):
 
 if __name__ == '__main__':
     try:
-        main()
+        res = main()
     except KeyboardInterrupt:
         exit(1)
     except Exception as e:
@@ -819,4 +822,4 @@ if __name__ == '__main__':
         exit(1, "virt-who failed: %s" % str(e))
     logger = logging.getLogger("virtwho.main")
     logger.debug("virt-who terminated")
-    exit(0)
+    exit(res)
