commit 4f375bb5693267eb9f6a783bd12c9a611d293a0d
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Tue Jan 12 07:38:05 2016 +0100

    Fix typo in template.conf
    
    Fixes #32.

diff --git a/template.conf b/template.conf
index 1778c9d..343c64e 100644
--- a/template.conf
+++ b/template.conf
@@ -12,7 +12,7 @@
 #server=             ; insert hostname or ip address of the server to connect to
 #username=           ; username for server authentication
 #password=           ; password for server authentication
-#encryped_password=  ; password encrypted using virt-who-password utility
+#encrypted_password=  ; password encrypted using virt-who-password utility
 #owner=              ; owner for use with SAM, Customer Portal, or Satellite 6
 #env=                ; environment for use with SAM, Customer Portal, or Satellite 6
 #hypervisor_id=      ; how will be the hypervisor identified, one of: uuid, hostname, hwuuid
@@ -25,7 +25,7 @@
 #server=
 #username=
 #password=
-#encryped_password=
+#encrypted_password=
 #owner=
 #env=
 #hypervisor_id=
