commit a8ff36d322192378dfa7e38b6592f86dbabb6511
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Thu May 14 13:46:42 2015 +0200

    Show error when encrypted password is corrupted

diff --git a/config.py b/config.py
index 4c7ab3b..9cd1689 100644
--- a/config.py
+++ b/config.py
@@ -34,6 +34,9 @@ class InvalidOption(Error):
     pass
 
 
+class InvalidPasswordFormat(Exception):
+    pass
+
 def parse_list(s):
     '''
     Parse comma-separated list of items that might be in double-quotes to the list of strings
@@ -96,6 +99,8 @@ class Config(object):
             try:
                 encrypted_password = parser.get(name, "encrypted_password")
                 password = Password.decrypt(unhexlify(encrypted_password))
+            except TypeError:
+                raise InvalidPasswordFormat("Password can't be decrypted, possibly corrupted")
             except NoOptionError:
                 password = None
 
diff --git a/virtwho.py b/virtwho.py
index bcb150c..0550872 100644
--- a/virtwho.py
+++ b/virtwho.py
@@ -30,7 +30,7 @@ import atexit
 from daemon import daemon
 from virt import Virt, DomainListReport, HostGuestAssociationReport, ErrorReport
 from manager import Manager, ManagerError, ManagerFatalError
-from config import Config, ConfigManager
+from config import Config, ConfigManager, InvalidPasswordFormat
 from password import InvalidKeyFile
 
 import log
@@ -522,7 +522,7 @@ def main():
     global virtWho
     try:
         virtWho = VirtWho(logger, options)
-    except InvalidKeyFile as e:
+    except (InvalidKeyFile, InvalidPasswordFormat) as e:
         logger.error(str(e))
         sys.exit(1)
 
