commit a0ec5992f79a25d36b55d96cf3e3b27a6322402f
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Tue May 26 09:22:32 2015 +0200

    log: don't rotate log file, it already has logrotate

diff --git a/log.py b/log.py
index 2745537..5be19b6 100644
--- a/log.py
+++ b/log.py
@@ -40,7 +40,7 @@ def getLogger(debug, background):
 
     # Try to write to /var/log, fallback on console logging:
     try:
-        fileHandler = logging.handlers.RotatingFileHandler(path, maxBytes=0x100000, backupCount=5)
+        fileHandler = logging.handlers.WatchedFileHandler(path)
         fileHandler.setFormatter(logging.Formatter('%(asctime)s [%(levelname)s]  @%(filename)s:%(lineno)d - %(message)s'))
         if debug:
             fileHandler.setLevel(logging.DEBUG)
