commit e580661d433fba774d0497254bd6b2476cf69b1e
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Thu Jun 4 16:21:42 2015 +0200

    subscriptionmanager: suppress BadStatusLine exception

diff --git a/manager/subscriptionmanager/subscriptionmanager.py b/manager/subscriptionmanager/subscriptionmanager.py
index 312040b..a77b210 100644
--- a/manager/subscriptionmanager/subscriptionmanager.py
+++ b/manager/subscriptionmanager/subscriptionmanager.py
@@ -19,6 +19,7 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 """
 
 import os
+from httplib import BadStatusLine
 
 import rhsm.connection as rhsm_connection
 import rhsm.certificate as rhsm_certificate
@@ -129,7 +130,10 @@ class SubscriptionManager(Manager):
         self._connect(**kwargs)
 
         # Send the mapping
-        return self.connection.hypervisorCheckIn(config.owner, config.env, mapping)
+        try:
+            return self.connection.hypervisorCheckIn(config.owner, config.env, mapping)
+        except BadStatusLine:
+            raise ManagerError("Communication with subscription manager interrupted")
 
     def uuid(self):
         """ Read consumer certificate and get consumer UUID from it. """
