commit eb5616183f2fc40b94438becb08a7a76aa11570f
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Tue May 5 14:51:37 2015 +0200

    Terminate background processes when main process exits

diff --git a/virt/virt.py b/virt/virt.py
index fe47f0a..57c2a86 100644
--- a/virt/virt.py
+++ b/virt/virt.py
@@ -125,6 +125,7 @@ class Virt(Process):
         self.logger = logger
         self.config = config
         super(Virt, self).__init__()
+        self.daemon = True
 
     @classmethod
     def fromConfig(cls, logger, config):
