commit 15fe618beac4c80c6b1d928c1bbb57d062ed0f02
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Thu Mar 10 16:43:54 2016 +0100

    Satellite: fix satellite5 env var usage

diff --git a/config.py b/config.py
index 281f4c4..a5d45e5 100644
--- a/config.py
+++ b/config.py
@@ -150,7 +150,7 @@ class GlobalConfig(GeneralConfig):
         'background': False,
         'configs': '',
         'reporter_id': util.generateReporterId(),
-        'smType': 'sam',
+        'smType': None,
         'interval': 60
     }
     LIST_OPTIONS = (
@@ -211,8 +211,10 @@ class Config(GeneralConfig):
         except KeyError:
             if 'sat_server' in self._options:
                 return 'satellite'
-            else:
+            elif 'rhsm_hostname' in self._options:
                 return 'sam'
+            else:
+                return None
 
     def checkOptions(self, logger):
         # Server option must be there for ESX, RHEVM, and HYPERV
@@ -261,7 +263,7 @@ class Config(GeneralConfig):
                     logger.warn("Option `owner` is not used in non-remote libvirt connection")
 
     @classmethod
-    def fromParser(self, name, parser, defaults=None):
+    def fromParser(cls, name, parser, defaults=None):
         options = {}
         for option in parser.options(name):
             options[option] = parser.get(name, option)
diff --git a/manager/manager.py b/manager/manager.py
index d71c0a7..d74e1b0 100644
--- a/manager/manager.py
+++ b/manager/manager.py
@@ -60,8 +60,10 @@ class Manager(object):
         assert subscriptionmanager
         assert satellite
 
+        smType = config.smType or options.smType or 'sam'
+
         for subcls in cls.__subclasses__():
-            if subcls.smType == config.smType:
+            if subcls.smType == smType:
                 return subcls(logger, options)
 
-        raise KeyError("Invalid config type: %s" % config.smType)
+        raise KeyError("Invalid config type: %s" % smType)
diff --git a/tests/test_config.py b/tests/test_config.py
index 5d4827a..767fe34 100644
--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -419,6 +419,7 @@ server=1.2.3.4
 username=admin
 password=password
 owner=root
+rhsm_hostname=abc
 """)
         self.assertRaises(InvalidOption, ConfigManager, self.logger, self.config_dir)
 
@@ -431,9 +432,11 @@ server=1.2.3.4
 username=admin
 password=password
 env=env
+rhsm_hostname=abc
 """)
         self.assertRaises(InvalidOption, ConfigManager, self.logger, self.config_dir)
 
+
 class TestGeneralConfig(TestBase):
     """
     A group of unittests to ensure correct functionality of GeneralConfig
diff --git a/tests/test_manager.py b/tests/test_manager.py
index 51d94ae..4f5f5af 100644
--- a/tests/test_manager.py
+++ b/tests/test_manager.py
@@ -83,6 +83,7 @@ class TestSubscriptionManager(TestManager):
     def test_sendVirtGuests(self, create_from_file, connection):
         self.prepare(create_from_file, connection)
         config = Config('test', 'libvirt')
+        config.smType = 'sam'
         manager = Manager.fromOptions(self.logger, self.options, config)
         manager.sendVirtGuests(self.domain_report, self.options)
         manager.connection.updateConsumer.assert_called_with(
@@ -94,6 +95,7 @@ class TestSubscriptionManager(TestManager):
     def test_hypervisorCheckIn(self, create_from_file, connection):
         self.prepare(create_from_file, connection)
         config = Config('test', 'libvirt')
+        config.smType = 'sam'
         manager = Manager.fromOptions(self.logger, self.options, config)
         self.options.env = "ENV"
         self.options.owner = "OWNER"
