commit 4ddbb759cace9c2c31aec021cdb157712d46b56c
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Tue Feb 23 14:27:22 2016 +0100

    Change pidfile permissions to 600

diff --git a/virtwho.py b/virtwho.py
index 4ecc3a7..852119c 100644
--- a/virtwho.py
+++ b/virtwho.py
@@ -651,8 +651,8 @@ class PIDLock(object):
 
     def is_locked(self):
         try:
-            f = open(self.filename, "r")
-            pid = int(f.read().strip())
+            with open(self.filename, "r") as f:
+                pid = int(f.read().strip())
             try:
                 os.kill(pid, 0)
                 return True
@@ -667,7 +667,7 @@ class PIDLock(object):
     def __enter__(self):
         # Write pid to pidfile
         try:
-            with open(self.filename, "w") as f:
+            with os.fdopen(os.open(self.filename, os.O_WRONLY | os.O_CREAT, 0600), 'w') as f:
                 f.write("%d" % os.getpid())
         except Exception as e:
             print >>sys.stderr, "Unable to create pid file: %s" % str(e)
