commit 6f8a2127f642811609157eaf5423047c088d1edf
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Thu May 14 13:48:49 2015 +0200

    rhevm: don't fail when hwuuid is not present

diff --git a/virt/rhevm/rhevm.py b/virt/rhevm/rhevm.py
index 244dbcf..d2d5f2a 100644
--- a/virt/rhevm/rhevm.py
+++ b/virt/rhevm/rhevm.py
@@ -84,7 +84,10 @@ class RhevM(virt.Virt):
             if self.config.hypervisor_id == 'uuid':
                 host_ids[id] = id
             elif self.config.hypervisor_id == 'hwuuid':
-                host_ids[id] = host.find('hardware_information').find('uuid').text
+                try:
+                    host_ids[id] = host.find('hardware_information').find('uuid').text
+                except AttributeError:
+                    self.logger.warn("Host %s doesn't have hardware uuid", id)
             elif self.config.hypervisor_id == 'hostname':
                 host_ids[id] = host.find('name').text
             else:
