commit 0c08e0a6500b6f3e42b961f79af9a063934b0b9e
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Thu May 14 13:49:13 2015 +0200

    limit queue size to number of virt backends

diff --git a/virtwho.py b/virtwho.py
index 0550872..a682506 100644
--- a/virtwho.py
+++ b/virtwho.py
@@ -151,7 +151,7 @@ class VirtWho(object):
 
         # Queue for getting events from virt backends
         if self.queue is None:
-            self.queue = Queue()
+            self.queue = Queue(len(self.configManager.configs))
 
         # Run the virtualization backends
         self.virts = []
