diff -up virt-who-0.16/virt/vdsm/vdsm.py.fix virt-who-0.16/virt/vdsm/vdsm.py
--- virt-who-0.16/virt/vdsm/vdsm.py.fix	2016-03-31 16:03:49.768185672 +0200
+++ virt-who-0.16/virt/vdsm/vdsm.py	2016-03-31 16:04:08.199034275 +0200
@@ -97,7 +97,7 @@ class Vdsm(Virt):
         CERTFILE = self.trust_store_path + '/certs/vdsmcert.pem'
         CACERT = self.trust_store_path + '/certs/cacert.pem'
 
-        ctx = SSL.Context('sslv3')
+        ctx = SSL.Context()
 
         ctx.set_verify(SSL.verify_peer | SSL.verify_fail_if_no_peer_cert, 16)
         ctx.load_verify_locations(CACERT)
