commit b9fc96493776e120ad8602effd7622c0cf35cc55
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Thu Jan 28 10:07:19 2016 +0100

    Background and oneshot modes are no longer exclusive

diff --git a/virtwho.py b/virtwho.py
index db2d338..a3642ec 100644
--- a/virtwho.py
+++ b/virtwho.py
@@ -499,8 +499,7 @@ def parseOptions():
 
     # Used only when starting as service (initscript sets it to 1, systemd to 0)
     env = os.getenv("VIRTWHO_BACKGROUND", "0").strip().lower()
-    if env in ["1", "true"]:
-        options.background = True
+    options.background = env in ["1", "true"]
 
     log.init(options)
     logger = log.getLogger(name='init', queue=False)
@@ -629,10 +628,6 @@ def parseOptions():
         if not options.env:
             raise OptionError("Option --%s-env (or VIRTWHO_%s_ENV environment variable) needs to be set" % (options.virtType, options.virtType.upper()))
 
-    if options.background and options.oneshot:
-        logger.error("Background and oneshot can't be used together, using background mode")
-        options.oneshot = False
-
     if options.interval < MinimumSendInterval:
         if not options.interval or options.interval == parser.defaults['interval']:
             logger.info("Interval set to the default of %d seconds.", DefaultInterval)
