commit 35c200153a35646d76799381dd8223742b051c9d
Author: Radek Novacek <rnovacek@redhat.com>
Date:   Tue Apr 21 15:27:17 2015 +0200

    Show info message that report was successful

diff --git a/virtwho.py b/virtwho.py
index d2db309..98d8956 100644
--- a/virtwho.py
+++ b/virtwho.py
@@ -119,6 +119,7 @@ class VirtWho(object):
     def _sendGuestList(self, report):
         manager = Manager.fromOptions(self.logger, self.options)
         manager.sendVirtGuests(report.guests)
+        self.logger.info("virt-who guest list update successful")
 
     def _sendGuestAssociation(self, report):
         manager = Manager.fromOptions(self.logger, self.options)
@@ -134,6 +135,9 @@ class VirtWho(object):
             guests = [x['guestId'] for x in created['guestIds']]
             self.logger.info("Created host: %s with guests: [%s]", created['uuid'], ", ".join(guests))
 
+        if not result['failedUpdate']:
+            self.logger.info("virt-who host/guest association update successful")
+
     def run(self):
         if not self.options.oneshot:
             self.logger.debug("Starting infinite loop with %d seconds interval" % self.options.interval)
